plugins/outputs/file/file.go

```diff
@@ -13,24 +14,86 @@ import (
 	"github.com/influxdata/telegraf/internal/rotate"
 	"github.com/influxdata/telegraf/plugins/outputs"
 	"github.com/influxdata/telegraf/plugins/serializers"
+	"github.com/klauspost/compress/zstd"
+	gzip "github.com/klauspost/pgzip"
 )
 
 //go:embed sample.conf
 var sampleConfig string
 
+var ValidCompressionAlgorithmLevels = map[string][]int{
+	"zstd": {1, 3, 7, 11},
+	"gzip": {-3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+}
+
 type File struct {
-	Files               []string        `toml:"files"`
-	RotationInterval    config.Duration `toml:"rotation_interval"`
-	RotationMaxSize     config.Size     `toml:"rotation_max_size"`
-	RotationMaxArchives int             `toml:"rotation_max_archives"`
-	UseBatchFormat      bool            `toml:"use_batch_format"`
-	Log                 telegraf.Logger `toml:"-"`
+	Files                []string        `toml:"files"`
+	RotationInterval     config.Duration `toml:"rotation_interval"`
+	RotationMaxSize      config.Size     `toml:"rotation_max_size"`
+	RotationMaxArchives  int             `toml:"rotation_max_archives"`
+	UseBatchFormat       bool            `toml:"use_batch_format"`
+	CompressionAlgorithm string          `toml:"compression_algorithm"`
+	CompressionLevel     int             `toml:"compression_level"`
+	encoder              interface{}
+	Log                  telegraf.Logger `toml:"-"`
 
 	writer     io.Writer
 	closers    []io.Closer
 	serializer serializers.Serializer
 }
 
+func validateCompressionAlgorithm(algorithm string) error {
+	for validAlgorithm := range ValidCompressionAlgorithmLevels {
+		if algorithm == validAlgorithm {
+			return nil
+		}
+	}
+	return fmt.Errorf("unknown or unsupported algorithm provided: %s", algorithm)
+}
+
+func validateCompressionLevel(algorithm string, level int) error {
+	for _, validAlgorithmLevel := range ValidCompressionAlgorithmLevels[algorithm] {
+		if level == validAlgorithmLevel {
+			return nil
+		}
+	}
+	return fmt.Errorf("unsupported compression level provided: %d. only %v are supported", level, ValidCompressionAlgorithmLevels[algorithm])
+}
+
+func CompressZstd(encoder *zstd.Encoder, src []byte) []byte {
+	return encoder.EncodeAll(src, make([]byte, 0, len(src)))
+}
+
+func CompressGzip(data []byte, level int) ([]byte, error) {
+	var b bytes.Buffer
+	gz,err := gzip.NewWriterLevel(&b, level)
```

Comment:
This function could work with io.Writer in the File struct but I couldn't get it working. I tried the whole night without success so I decided to get it working without touching the File struct.

Reply:
I knew why it didn't work. basically I didn't want to recreate a new gzip writer every time I want to write so that means I never called gzip.Close() which is bad and caused broken gzip files to be created with unexpected EOF errors.